src/wp-includes/global-styles-and-settings.php

```diff
@@ -306,8 +306,30 @@ function wp_add_global_styles_for_blocks() {
 
 	$tree        = WP_Theme_JSON_Resolver::get_merged_data();
 	$block_nodes = $tree->get_styles_block_nodes();
+
+	$can_use_cached = ! wp_is_development_mode( 'theme' );
+	if ( $can_use_cached ) {
+		$settings_hash = _wp_get_global_settings_hash();
+		$cache_key     = 'wp_styles_for_blocks';
```

Comment:
Why not use the `$settings_hash` as a part of this cache key instead of adding it to each style cache?

```suggestion
		$cache_key     = "wp_styles_for_blocks:$settings_hash";
```

Reply:
It was due to the delete transition. I wanted a label that was easy to delete. But, yes I can do this and add a pattern based delete in delete transition.

Reply:
Ah, that makes sense. Given that we are implementing this as a transient with an expiration time, I don't think we need a delete transition at all. If the `$settings_hash` changes, it will naturally cause a cache miss and a new transient will be created. The old transient will naturally expire and get cleaned up when WP deletes expired transients.

Reply:
That makes sense. If something changes it will change the settings too in most cases. I was wondering, Will there be cases where settings won't change but the results would like an update of the theme?

Reply:
That's a good question. I think the assumption being made in your PR so far is that the combination of the hashed global settings object in combination with the hashed block metadata for each block, is enough of a fingerprint to use for cache busting. I think this is probably correct, but it occurred to me that rather than hashing each block node individually, we could also hash the full `$block_nodes` object as part of the transient key and avoid hashing during each instance of the loop, since changes to any blocks would invalidate the whole cache. I've demonstrated this and some of my other suggestions in the following diff:

```diff
diff --git a/src/wp-includes/global-styles-and-settings.php b/src/wp-includes/global-styles-and-settings.php
index 639c99b409..901dc95e68 100644
--- a/src/wp-includes/global-styles-and-settings.php
+++ b/src/wp-includes/global-styles-and-settings.php
@@ -309,23 +309,26 @@ function wp_add_global_styles_for_blocks() {
 
 	$can_use_cached = ! wp_is_development_mode( 'theme' );
 	if ( $can_use_cached ) {
+		$block_hash    = md5( wp_json_encode( $block_nodes ) );
 		$settings_hash = _wp_get_global_settings_hash();
-		$cache_key     = 'wp_styles_for_blocks';
+		$cache_key     = "wp_styles_for_blocks:$block_hash:$settings_hash";
 		$cached        = get_site_transient( $cache_key );
 		if ( ! is_array( $cached ) ) {
 			$cached = array();
 		}
 	}
 
+	$update_cache = false;
+
 	foreach ( $block_nodes as $metadata ) {
 		if ( $can_use_cached ) {
-			$metadata_hash   = md5( wp_json_encode( $metadata ) );
-			$style_cache_key = $settings_hash . '_' . $metadata_hash;
-			if ( isset( $cached[ $style_cache_key ] ) ) {
-				$block_css = $cached[ $style_cache_key ];
+			$block_name = str_replace( '/', '-', $metadata['name'] );
+			if ( isset( $cached[ $block_name ] ) ) {
+				$block_css = $cached[ $block_name ];
 			} else {
-				$block_css                  = $tree->get_styles_for_block( $metadata );
-				$cached[ $style_cache_key ] = $block_css;
+				$block_css             = $tree->get_styles_for_block( $metadata );
+				$cached[ $block_name ] = $block_css;
+				$update_cache          = true;
 			}
 		} else {
 			$block_css = $tree->get_styles_for_block( $metadata );
@@ -351,7 +354,7 @@ function wp_add_global_styles_for_blocks() {
 			if ( str_starts_with( $metadata['name'], 'core/' ) ) {
 				$block_name   = str_replace( 'core/', '', $metadata['name'] );
 				$block_handle = 'wp-block-' . $block_name;
-				if ( in_array( $block_handle, $wp_styles->queue ) ) {
+				if ( in_array( $block_handle, $wp_styles->queue, true ) ) {
 					wp_add_inline_style( $stylesheet_handle, $block_css );
 				}
 			} else {
@@ -366,7 +369,7 @@ function wp_add_global_styles_for_blocks() {
 				if ( str_starts_with( $block_name, 'core/' ) ) {
 					$block_name   = str_replace( 'core/', '', $block_name );
 					$block_handle = 'wp-block-' . $block_name;
-					if ( in_array( $block_handle, $wp_styles->queue ) ) {
+					if ( in_array( $block_handle, $wp_styles->queue, true ) ) {
 						wp_add_inline_style( $stylesheet_handle, $block_css );
 					}
 				} else {
@@ -376,7 +379,7 @@ function wp_add_global_styles_for_blocks() {
 		}
 	}
 
-	if ( $can_use_cached ) {
+	if ( $update_cache ) {
 		set_site_transient( $cache_key, $cached, HOUR_IN_SECONDS );
 	}
 }

```

Reply:
@joemcgill While this code should have worked, strangely, some `$metadata` doesn't contain a block name, causing the code to fail. 
This is one sample array I got, 
```
Array
(
    [path] => Array
        (
            [0] => styles
            [1] => blocks
            [2] => core/navigation
            [3] => elements
            [4] => link
        )

    [selector] => .wp-block-navigation a:where(:not(.wp-element-button))
)
```

I am checking to see if we can skip those blocks entirely from logic. It might make the code more performant.